cli/src/utils/type_example.rs

```diff
@@ -0,0 +1,380 @@
+use color_eyre::eyre::eyre;
+
+use scale_info::{
+    form::PortableForm, Field, PortableRegistry, TypeDef, TypeDefArray, TypeDefPrimitive,
+    TypeDefTuple, TypeDefVariant,
+};
+use scale_value::{Value, ValueDef};
+use std::fmt::Write;
+use std::write;
+
+pub fn print_type_examples<T>(
+    ty: &T,
+    registry: &PortableRegistry,
+    type_placeholder: &str,
+) -> color_eyre::Result<String>
+where
+    T: TypeExample,
+{
+    let type_examples = ty.type_example(registry)?;
+    let mut output = String::new();
+    match type_examples.len() {
+        0 => {
+            write!(
+                output,
+                "There are no examples available for a {type_placeholder} matching this shape:"
+            )?;
+        }
+        1 => {
+            write!(
+                output,
+                "Here is an example of a {type_placeholder} matching this shape:"
+            )?;
+        }
+        i => {
+            write!(
+                output,
+                "Here are {i} examples of a {type_placeholder} matching this shape:"
+            )?;
+        }
+    };
+    for self_value in type_examples {
+        let value = <T as TypeExample>::upcast(self_value);
+        let example_str = scale_value::stringify::to_string(&value);
+        write!(output, "\n{}", example_str)?;
+    }
+    Ok(output)
+}
+
+/// a trait for producing scale value examples for a type.
+pub trait TypeExample {
```

Comment:
I think this is ok for now, but I'm not super convinced that a trait is necessary here (or above). I'd think that in general if you want to give an example of some type, you could expose a single function, `type_example(type_id: u32, registry: &PortableRegistry, indent: 32) -> String`  that would be the entire interface to this stuff, which would make it quite easy to change around the internals (eg stop relying on scale-value) or whatever without worrying about other code breaking. Then, you'd just have internal functions to handle the different types rather than trait impls.

Reply:
That is true, maybe I don't need to expose the trait. It just made it easier to reason about the examples for me. I can clean this up.